test: add cppapi test to give TorchCodecConfig.cmake a try

dvrogozh opened the pull request:
That's a test for the `TorchCodecConfig.cmake` we've introduced in https://github.com/meta-pytorch/torchcodec/commit/dc872282ed7e019422214a1e53681a87d7b37f05. Test is integrated into TorchCodec pytest suite and marked with `@pytest.mark.devel` which is disabled by default. Test assumes that ffmpeg is installed and can be used for both runtime and compilation (i.e. .pc and include files are available). Test builds a .so which is linked to TorchCodec core library. There are 2 kind of tests executed:
1. Building with ffmpeg being found via `pkg-config`
2. Building with ffmpeg being pointed to by `TORCHCODEC_FFMPEGn_INSTALL_PREFIX` environment variable

Test which relies on environment variable assumes that the ffmpeg install layout will match expectations hardcoded in the test and in the `TorchCodecConfig.cmake`. It potentially might fail - we need to check in the CI.

@NicolasHug, please, help to review paying attention on the above. I hope this will allow us to have test naturally integrated into TorchCodec test suite. If not, let's discuss what we need to modify.

CC: @NicolasHug

dvrogozh commented:
> is that reasonable to request having CUDA fully installed on the runner environment 

Actually, it's installed, but was not detected. I guess that due to OS differences in cmake behavior. On ubuntu it get's detected from /usr/local, on AlmaLinux it seems not:
```
$ docker run -it --rm pytorch/manylinux2_28-builder:cuda12.6 ls /usr/local/cuda/include
bootstrap_device_host        cudaTypedefs.h                 cusparseLt.h                              nv_decode.h
builtin_types.h              cudaVDPAU.h                    cusparse_v2.h                             nvFatbin.h
```

NicolasHug commented:
Thanks for the PR @dvrogozh ! I have a few questions / comments:

- Do we want all this to work when users are building torchcodec from source, or do we want it to work when they install a pre-built version of torchcodec (e.g. via `pip install torchcodec ...`) ? I assume we want the latter?
- Would it be acceptable to only support Linux for now? I feel like Windows and MacOS support might be lower priority for the feature, and focusing on linux could make it easier.

> we need not only runtime, but development environment as well, i.e. we need pkg-config, CUDA dev. files and ffmpeg dev. files available.

Right now, the newly added test are ran under this `install-and-test` job: https://github.com/meta-pytorch/torchcodec/blob/38fa96c6226ce502c0f86fffa664835059d35ce9/.github/workflows/linux_wheel.yaml#L60. The point of this job is to have a minimal environment that looks like what an end-user would have (typically without development dependencies!). If we need dev stuff, it's best to run the new test in a separate job. That is, a new job, similar in nature to `install-and-test`, that still depends on the `build` job above, but where we can install the required dev dependencies safely. This new job doesn't even need to rely on pytest, we could have some inline bash commands directly calling into `cmake`, which might be simpler too? Does that make sense?

dvrogozh commented:
@NicolasHug, I reworked the test trying to take into account your comments. Key points:
1. Added new tests marked with `@pytest.mark.devel`
2. Tests are skipped by default and require explicit `-m devel` on the command line
3. Added new test job only for linux/torch-cpu to give test a run

Please, help to review.
